helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java

```diff
@@ -0,0 +1,132 @@
+package org.apache.helix.controller.changedetector;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import org.apache.helix.HelixConstants;
+import org.apache.helix.HelixException;
+import org.apache.helix.HelixProperty;
+import org.apache.helix.controller.dataproviders.ResourceControllerDataProvider;
+
+public class ResourceChangeDetector implements ChangeDetector {
+
+  private ResourceChangeCache _oldCache; // cache snapshot for previous pipeline run
```

Comment:
Interesting idea - could you explain a little further how keeping multiple caches would be useful? That sounds like the opposite direction we're heading (since we want to reduce the memory footprint).

Reply:
I am trying to throw some ideas. For example, for debugging purpose, we would like to enable more than 1 round of assignment. Or some fancy AI algorithm requires more empirical data. You can leave the room for this kind of potential usage. Yes, we definitely need to reduce the memory usage. But we can leave some flexibilities here.

I will leave to you to make the decision.

Reply:
I see where you're coming from. It sounds like the logic would look quite different in that case than what we want here. I think in that case, we could always add more implementations of ChangeDetector as we make further progress that direction.
For the sake of keeping it simple and not over-engineering, I'll keep this one as is (as in supporting 2 rounds of rebalances).